Removed unnecessary unsafe fns.

ObsceneGiraffe opened the pull request:
- add warn unsafe_code lint.

Is there a use-case for having the unsafe functions? 
The use of `unsafe` keyword here is stretching it's intended use. 
Using `unsafe` is intended to turn off memory safety checks.
But `prae` is using it to mean "turn off data integrity checks"

teenjuna commented:
These functions are handy for performance  optimisations. If the user has a very frequent operation and can guarantee that the invariants of provided data are upheld, he can avoid the overhead of calling `adjust` and `mutate`. This was part of [this](https://www.reddit.com/r/rust/comments/occg7a/prae_a_simple_library_that_helps_you_keep_your/h3uvob0?utm_source=share&utm_medium=web2x&context=3) discussion. 

But the use of `unsafe` might violate its intended usage, as was also discussed in the discussion linked.  Given that some crate might have `forbid(unsafe_code)` and the functions clearly say that one must be cautious  using them, we can just remove `unsafe` keywords from them.

teenjuna commented:
Perhaps they should be called `_unchecked`, but I'm not sure if it's clear that the functions not only don't call `validate`, but also don't call `adjust`. What do you think?

ObsceneGiraffe commented:
I can see someone wanting to use unchecked versions of the functions for optimization reasons. I would prefer that was behind a feature gate. As in, if I was a lead on a project and `prae` was included, I may want to disable the possibility of unchecked mutability.

Instead of removing the unchecked functions, let's move them into a separate trait that is behind a feature gate then if someone consciously chooses to use unchecked mutability, they need to bring the trait into scope.

teenjuna commented:
Seems like a good solution. We can call the feature `manual_access`. Do you want to do that?

ObsceneGiraffe commented:
Yea I'll implement that :wink: I'm attending an all day birthday shindig today. But I'll take a break from it all at some stage and see what I can get done.

teenjuna commented:
Super :)

ObsceneGiraffe commented:
I'd go with the name `unchecked_access` as that maps to the function names. `manual` implies that something was automatic and it now done by hand, which is a smooth fit in this case.

ObsceneGiraffe commented:
I added a new feature `unchecked-access` and an associated trait `UncheckedAccess` that contains the unchecked fns.
I also added tests for the unchecked fns.

teenjuna commented on prae/src/core.rs:
```diff
@@ -101,32 +101,50 @@ where
 
+    /// Verifies invariants. This is guaranteed to succeed unless you've used
+    /// one of the `unsafe` methods that require variants to be manually upheld.
+    pub fn verify(&self) -> Result<(), E> {
```
Let's move it to the `UncheckedAccess`

ObsceneGiraffe replied:
Done

teenjuna commented on prae/src/core.rs:
```diff
@@ -101,32 +101,50 @@ where
+impl UncheckedAccess for Guard {
+    /// Construct a value without calling `adjust` and `validate`. The invariant must be upheld
+    /// manually. Should be used only for optimisation purposes.
+    unsafe fn new_unchecked<V: Into<T>>(v: V) -> Self {
```
I thought we agreed that `unsafe` should not be used?

ObsceneGiraffe replied:
All removed now.

teenjuna commented on prae/tests/unchecked.rs:
```diff
@@ -0,0 +1,54 @@
+
+    assert_matches!(
+        Username::new("").unwrap_err(),
+        prae::ValidationError { .. }
```
Oops, there is no `ValidationError` anymore!

ObsceneGiraffe replied:
I've merge in upstream master, but the build is failing due to the serde_integration mod being compiled even when the serde feature is disabled. See pull request  #19 .

teenjuna commented on prae/src/core.rs:
```diff
@@ -101,32 +101,50 @@ where
+}
+
+#[cfg(feature = "unchecked-access")]
+trait UncheckedAccess {
```
Now I wonder if there's a need for a trait. Seems like adding another `impl` block guarded by the feature gate is enough. This code works fine:
```rust
struct S;

impl S {
    fn a() {}
}

impl S {
    fn b() {}
}

fn main() {
    S::a();
    S::b();
}
```

teenjuna replied:
It will be just much cleaner in my opinion

ObsceneGiraffe replied:
My thinking here was that given the unchecked features are quite specialist, I want the user to explicitly import them into the scope they want to use them in. Maybe it's a little over the top? I was imagining the case where I want to know exactly when in my source code the unchecked functions are used. But this can easily be done with a *_unchecked search across the code base.

Let's go with simplicity here, I'll remove the trait.

ObsceneGiraffe replied:
I've moved the unchecked fns into the impl and deleted the `UncheckedAccess` trait.

teenjuna commented on prae/tests/unchecked.rs:
```diff
@@ -0,0 +1,54 @@
+
+    let mut un = Username::new("user").unwrap();
+    assert_matches!(
+        un.try_mutate(|u| *u = "".to_owned()).unwrap_err(),
```
This is more of a matter of taste, but let's not call `.unwrap_err()` in `assert_matches!` since it's more consistent with the rest of the tests, a bit shorter and gives a better error in case the method returns `Ok(...)`

ObsceneGiraffe replied:
Done

teenjuna commented on prae/tests/unchecked.rs:
```diff
@@ -0,0 +1,54 @@
+    let t = un.get_mut();
+    t = "";
+    assert_eq!(un.get(), "");
+}
```
Looks like you didn't run `cargo fmt`

ObsceneGiraffe replied:
There were many errors in the tests, they're passing now :wink:

teenjuna commented on prae/src/core.rs:
```diff
@@ -101,32 +101,50 @@ where
     }
 
+    /// Verifies invariants. This is guaranteed to succeed unless you've used
+    /// one of the `unsafe` methods that require variants to be manually upheld.
```
```suggestion
    /// one of the `_unchecked` methods that require variants to be manually upheld.
```

ObsceneGiraffe replied:
Done

teenjuna reviewed (changes requested):
Thanks for your help, here are some comments

teenjuna commented:
Thank you!
